File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
def bfs(start,graph,visited):
    count = 1
    queue = deque([start])
    visited[start] = 1
    
    while queue:
        node = queue.popleft()
        for next_node in graph[node]:
            if not visited[next_node]:
                visited[next_node] = 1
                queue.append(next_node)
                count += 1
    
    return count
def solution(n, wires):
    answer = n
    for i in range(len(wires)):
        graph = [[] for _ in range(n+1)]
        for j in range(len(wires)):
            if i == j:
                continue
            a, b = wires[j]
            graph[a].append(b)
            graph[b].append(a)
        visited = [0] * (n+1)
        group_count = bfs(1,graph,visited)
        diff = abs((n - (group_count)) - group_count)
        answer = min(answer,diff)

    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {

        int answer = Integer.MAX_VALUE; 

        
        for (int i = 0 ; i < wires.length; i++) {
            int[][] temp = new int[wires.length-1][2]; 
            
            if(i>0) System.arraycopy(wires, 0, temp, 0, i);
            
            if(i<wires.length-1) System.arraycopy(wires, i+1, temp, i, wires.length - i - 1);
            
            List<List<Integer>> graph = new ArrayList<>();
            for (int j = 1; j < n+2; j++) {
                graph.add(new ArrayList<>());
            }
            
            for (int[] wire : temp) {
                graph.get(wire[0]).add(wire[1]);
                graph.get(wire[1]).add(wire[0]);
            }
            
            int res = bfs(1,graph,n);
            int diff = Math.abs(res - (n - res));
            answer = Math.min(answer, diff);
            
          
        }
        return answer;
    }
    
    private int bfs(int start,List<List<Integer>> graph,int n) {
        
        List<Integer> visited = new ArrayList<>();
        for (int i = 0; i < n+1 ; i++) {
            visited.add(0);
        }
        
        visited.set(start, 1);
        Deque<Integer> queue = new ArrayDeque<>();
        queue.add(start);
        int cnt = 1;
        
        while (!queue.isEmpty()) {
            int node = queue.poll();
            
            for (int nextNode : graph.get(node)) {
                if (visited.get(nextNode) == 0) {
                    visited.set(nextNode,1);
                    queue.offer(nextNode);
                    cnt ++;
                }
            }
        }
        
        return cnt;
        
    }
    
    
}